From eb41c09172f3bc92731aec62caf4b215cf6d9873 Mon Sep 17 00:00:00 2001
From: Valerio Valerio <valerio.valerio@jollamobile.com>
Date: Tue, 3 Feb 2015 11:22:59 +0200
Subject: [PATCH 14/20] Use periodic keepalive instead of a long running
 one(IMAP IDLE).

This is a partial revert of d45bc6ca22e5f7e7a6b8e4e38ebad29794f61fcd
where a long running keepalive has introduced, but periodic is proven
to cause less battery comsumption in the long run.
---
 .../messageservices/imap/imapclient.cpp       |  5 ++---
 .../messageservices/imap/imapservice.cpp      | 20 +++++++++++++++++--
 .../messageservices/imap/imapservice.h        |  2 ++
 3 files changed, 22 insertions(+), 5 deletions(-)

diff --git a/src/plugins/messageservices/imap/imapclient.cpp b/src/plugins/messageservices/imap/imapclient.cpp
index fec1d06d..1840fa79 100644
--- a/src/plugins/messageservices/imap/imapclient.cpp
+++ b/src/plugins/messageservices/imap/imapclient.cpp
@@ -1922,7 +1922,7 @@ bool ImapClient::pushEnabled()
 
 void ImapClient::setPushEnabled(bool state)
 {
-    if (_pushEnabled != state) {
+    if (pushEnabled() != state) {
         qMailLog(Messaging) << Q_FUNC_INFO << "Setting push enabled state to " << state;
         _pushEnabled = state;
     }
@@ -1984,8 +1984,7 @@ void ImapClient::onSsoSessionResponse(const QMap<QString,QList<QByteArray> > &ss
     }
     if (_sendLogin) {
         _protocol.sendLogin(_config, _ssoLogin);
-    }
-    if (_waitingForIdle) {
+    } else if (_waitingForIdle) {
         monitor(_waitingForIdleFolderIds);
     }
 }
diff --git a/src/plugins/messageservices/imap/imapservice.cpp b/src/plugins/messageservices/imap/imapservice.cpp
index 3dae8ed1..314052bc 100644
--- a/src/plugins/messageservices/imap/imapservice.cpp
+++ b/src/plugins/messageservices/imap/imapservice.cpp
@@ -1534,8 +1534,10 @@ ImapService::ImapService(const QMailAccountId &accountId)
     connect(_initiatePushEmailTimer, SIGNAL(timeout()), this, SLOT(initiatePushEmail()));
 
 #ifdef USE_KEEPALIVE
+     _idling = false;
     _backgroundActivity = new BackgroundActivity(this);
     _backgroundActivity->setWakeupFrequency(BackgroundActivity::ThirtySeconds);
+    connect(_backgroundActivity, SIGNAL(running()), this, SLOT(onUpdateLastSyncTime()));
 
     // Connect to dbus signals emitted by buteo notifying schedule changes
     QDBusConnection m_dBusConnection(QDBusConnection::sessionBus());
@@ -1639,6 +1641,7 @@ void ImapService::disable()
     _restartPushEmailTimer->stop();
     _initiatePushEmailTimer->stop();
 #ifdef USE_KEEPALIVE
+    _idling = false;
     _backgroundActivity->stop();
 #endif
     _source->setIntervalTimer(0);
@@ -1776,6 +1779,7 @@ void ImapService::initiatePushEmail()
     _restartPushEmailTimer->stop();
     _initiatePushEmailTimer->stop();
 #ifdef USE_KEEPALIVE
+     _idling = false;
     if (_backgroundActivity->isRunning()) {
         _backgroundActivity->stop();
         qMailLog(Messaging) << Q_FUNC_INFO <<  "Stopping keepalive";
@@ -1792,7 +1796,8 @@ void ImapService::initiatePushEmail()
         _establishingPushEmail = true;
 #ifdef USE_KEEPALIVE
         qMailLog(Messaging) << Q_FUNC_INFO <<  "Starting keepalive";
-        _backgroundActivity->run();
+         _idling = true;
+        _backgroundActivity->wait();
 #endif
         QMailAccount account(_accountId);
         const bool hasPersistentConnection = (account.status() & QMailAccount::HasPersistentConnection);
@@ -1853,11 +1858,22 @@ void ImapService::updateStatus(const QString &text)
 }
 
 #ifdef USE_KEEPALIVE
+void ImapService::onUpdateLastSyncTime()
+{
+    // start timer again if still in idle mode
+    if (_idling) {
+        _backgroundActivity->wait();
+    } else if (_backgroundActivity->isRunning()){
+        qMailLog(Messaging) << Q_FUNC_INFO << "Stopping keepalive";
+        _backgroundActivity->stop();
+    }
+}
+
 void ImapService::stopPushEmail()
 {
     qMailLog(Messaging) << "Stopping push email for account" << _accountId
                         << QMailAccount(_accountId).name();
-
+    _idling = false;
     QMailAccount account(_accountId);
     const bool hasPersistentConnection = (account.status() & QMailAccount::HasPersistentConnection);
     if (hasPersistentConnection) {
diff --git a/src/plugins/messageservices/imap/imapservice.h b/src/plugins/messageservices/imap/imapservice.h
index 856fb4d8..8e3d8ff2 100644
--- a/src/plugins/messageservices/imap/imapservice.h
+++ b/src/plugins/messageservices/imap/imapservice.h
@@ -71,6 +71,7 @@ protected slots:
     void updateStatus(const QString& text);
 
 #ifdef USE_KEEPALIVE
+    void onUpdateLastSyncTime();
     void stopPushEmail();
     void pushEnabledStatus(uint accountId, const QString &profileType, bool state);
 #endif
@@ -94,6 +95,7 @@ private:
     QTimer *_initiatePushEmailTimer;
 #ifdef USE_KEEPALIVE
     BackgroundActivity* _backgroundActivity;
+    bool _idling;
     bool _accountPushEnabled;
     bool _buteoReplyReceived;
 #endif
-- 
2.17.1

